File: src/module/role/role_protocol.erl
-module(role_protocol).
-export([read/2, write/2]).
-include("role.hrl").
-include("asset.hrl").
-include("vip.hrl").


read(10101, <<>>) ->
    {ok, []};

read(10102, <<>>) ->
    {ok, []};

read(10103, <<>>) ->
    {ok, []};

read(Code, Binary) ->
    {error, Code, Binary}.


write(10101, #role{role_id = RoleId, role_name = RoleName, sex = Sex, classes = Classes, level = Level, item_size = ItemSize, bag_size = BagSize, store_size = StoreSize}) ->
    {ok, protocol:pack(10101, <<RoleId:64, (byte_size(RoleName)):16, (RoleName)/binary, Sex:8, Classes:8, Level:64, ItemSize:16, BagSize:16, StoreSize:16>>)};

write(10102, #asset{gold = Gold, silver = Silver, copper = Copper, exp = Exp}) ->
    {ok, protocol:pack(10102, <<Gold:64, Silver:32, Copper:64, Exp:64>>)};

write(10103, #vip{vip_level = VipLevel, exp = Exp, expire_time = ExpireTime}) ->
    {ok, protocol:pack(10103, <<VipLevel:8, Exp:64, ExpireTime:32>>)};

write(Code, Content) ->
    {error, Code, Content}.

File: script/make/protocol/lua/meta/RoleProtocol.lua
roleProtocol = {
    [10101] = {
        ["comment"] = "角色",
        ["write"] = {},
        ["read"] = {
            {name = "roleId", type = "u64", comment = "角色ID", explain = {}},
            {name = "roleName", type = "bst", comment = "角色名", explain = {}},
            {name = "sex", type = "u8", comment = "性别", explain = {}},
            {name = "classes", type = "u8", comment = "职业", explain = {}},
            {name = "level", type = "u64", comment = "等级", explain = {}},
            {name = "itemSize", type = "u16", comment = "普通背包大小", explain = {}},
            {name = "bagSize", type = "u16", comment = "装备背包大小", explain = {}},
            {name = "storeSize", type = "u16", comment = "仓库背包大小", explain = {}}
        }
    },
    [10102] = {
        ["comment"] = "资产",
        ["write"] = {},
        ["read"] = {
            {name = "gold", type = "u64", comment = "金币", explain = {}},
            {name = "silver", type = "u32", comment = "银币", explain = {}},
            {name = "copper", type = "u64", comment = "铜币", explain = {}},
            {name = "exp", type = "u64", comment = "经验", explain = {}}
        }
    },
    [10103] = {
        ["comment"] = "vip",
        ["write"] = {},
        ["read"] = {
            {name = "vipLevel", type = "u8", comment = "等级", explain = {}},
            {name = "exp", type = "u64", comment = "经验", explain = {}},
            {name = "expireTime", type = "u32", comment = "过期时间", explain = {}}
        }
    }
}